crates/epaint/src/mutex.rs

```diff
@@ -146,51 +146,176 @@ mod rw_lock_impl {
+    pub struct RwLockWriteGuard<'a, T> {
+        // The option is used only because we need to `take()` the guard out of self
+        // when doing remappings (`map()`), i.e. it's used as a safe `ManuallyDrop`.
+        guard: Option<MappedRwLockWriteGuard<'a, T>>,
+        holders: Arc<parking_lot::Mutex<HashMap<ThreadId, backtrace::Backtrace>>>,
+    }
+
+    impl<'a, T> RwLockWriteGuard<'a, T> {
+        #[inline]
+        pub fn map<U, F>(mut s: Self, f: F) -> RwLockWriteGuard<'a, U>
+        where
+            F: FnOnce(&mut T) -> &mut U,
+        {
+            RwLockWriteGuard {
+                guard: s
+                    .guard
+                    .take()
+                    .map(|g| parking_lot::MappedRwLockWriteGuard::map(g, f)),
+                holders: Arc::clone(&s.holders),
+            }
+        }
+    }
+
+    impl<'a, T> Deref for RwLockWriteGuard<'a, T> {
+        type Target = T;
+
+        fn deref(&self) -> &Self::Target {
+            &**self.guard.as_ref().unwrap()
+        }
+    }
+
+    impl<'a, T> DerefMut for RwLockWriteGuard<'a, T> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            &mut **self.guard.as_mut().unwrap()
+        }
+    }
+
+    impl<'a, T> Drop for RwLockWriteGuard<'a, T> {
+        fn drop(&mut self) {
+            let tid = std::thread::current().id();
+            self.holders.lock().remove(&tid);
+        }
+    }
 
     /// Provides interior mutability.
     ///
     /// Uses `parking_lot` crate on native targets, and `atomic_refcell` on `wasm32` targets.
     #[derive(Default)]
     pub struct RwLock<T> {
         lock: parking_lot::RwLock<T>,
-        last_lock: parking_lot::Mutex<backtrace::Backtrace>,
+        // Technically we'd need a list of backtraces per thread-id since parking_lot's
+        // read-locks are reentrant.
+        // In practice it's not that useful to have the whole list though, so we only
+        // keep track of the first backtrace for now.
+        holders: Arc<parking_lot::Mutex<HashMap<ThreadId, backtrace::Backtrace>>>,
     }
 
     impl<T> RwLock<T> {
         pub fn new(val: T) -> Self {
             Self {
                 lock: parking_lot::RwLock::new(val),
-                last_lock: Default::default(),
+                holders: Default::default(),
             }
         }
 
         pub fn read(&self) -> RwLockReadGuard<'_, T> {
+            let tid = std::thread::current().id();
+
             assert!(
-                !self.lock.is_locked_exclusive(),
-                "{} DEAD-LOCK DETECTED! Previous lock held at:\n{}\n\n",
+                // There are two ways we can get out of this situation without deadlocking:
+                // - either the lock is not currently exclusively held by anyone,
+                // - or the current thread is not already holding the lock in any way
+                //   (read reentrancy).
+                !self.lock.is_locked_exclusive() || !self.holders.lock().contains_key(&tid),
```

Comment:
Maybe it's just me, but this boolean expression hurts my soft head 🙃

If we negate it we can give it a name:

```
// If it is write-locked, and we locked it (re-entrancy deadlocks):
let would_deadlock = self.lock.is_locked_exclusive() && self.holders.lock().contains_key(&tid);
assert!(!would_deadlock, …)`
```

same below